Parsing Memo Incomplete

Hi All

I'm getting an error when trying to write and then read a memo field.

Write call seems to succeed with no error thrown:


	newNote := Note{
		Text: "zHdVIxu2TuBmm/nY9CDbLotel+ToRIxApggFvFSCjsvU//uZXDjbXBfliSpkSubmzu/Imn7NTs6Y1zD7Q2rZHg==",
	}

	newRow, err := table.RowFromStruct(newD)
	if err != nil {
		panic(dbase.GetErrorTrace(err))
	}

	Add the new row to the database table.
	err = newRow.Write()
	if err != nil {
		panic(dbase.GetErrorTrace(err))
	}


But when I try to read the row I get:

panic: dbase-table-bytestorow-3:dbase-interpreter-parsememo-1:dbase-io-readmemo-5:parsing memo failed at column field: TEXT failed with error: INCOMPLETE 

Here's my read code:

	row, err := table.Row()
	if err != nil {
		panic(dbase.GetErrorTrace(err))
	}

Full log

[dbase] [DEBUG] 2022/10/27 17:28:35 Reading row: 21 at offset: 1139
[dbase] [DEBUG] 2022/10/27 17:28:35 Converting row data (31 bytes) to row struct...
[dbase] [DEBUG] 2022/10/27 17:28:35 Reading memo block 367 at position 23488
[dbase] [DEBUG] 2022/10/27 17:28:35 Memo block header => text: true, length: 88
[dbase] [ERROR] 2022/10/27 17:28:35 dbase-io-readmemo-5:INCOMPLETE
[dbase] [ERROR] 2022/10/27 17:28:35 dbase-interpreter-parsememo-1:parsing memo failed at column field: TEXT failed with error: INCOMPLETE
[dbase] [ERROR] 2022/10/27 17:28:35 dbase-table-bytestorow-3:parsing memo failed at column field: TEXT failed with error: INCOMPLETE
[dbase] [DEBUG] 2022/10/27 17:28:35 Closing file: ./files/AAQBX.DBF
[dbase] [DEBUG] 2022/10/27 17:28:35 Closing related file: ./files/AAQBX.DBF
panic: dbase-table-bytestorow-3:dbase-interpreter-parsememo-1:dbase-io-readmemo-5:parsing memo failed at column field: TEXT failed with error: INCOMPLETE


I noticed that this only happens when "Text" is greater than 56 characters.

Thanks for all the work you guys do.

--- a/dbase/io_unix.go
+++ b/dbase/io_unix.go
@@ -539,13 +539,17 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 	}
 	// Get the block position
 	blockPosition := file.memoHeader.NextFree
+	blocks := length / int(file.memoHeader.BlockSize)
+	if length%int(file.memoHeader.BlockSize) > 0 {
+		blocks++
+	}
 	// Write the memo header
-	err := file.writeMemoHeader()
+	err := file.writeMemoHeader(blocks)
 	if err != nil {
 		return nil, newError("dbase-io-writememo-2", err)
 	}
 	// Put the block data together
-	block := make([]byte, file.memoHeader.BlockSize)
+	block := make([]byte, length+8)
 	// The first 4 bytes are the signature, 1 for text, 0 for binary(image)
 	if text {
 		binary.BigEndian.PutUint32(block[:4], 1)
@@ -603,7 +607,8 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 }
 
 // writeMemoHeader writes the memo header to the memo file.
-func (file *File) writeMemoHeader() (err error) {
+// Size is the number of blocks the new memo data will take up.
+func (file *File) writeMemoHeader(size int) (err error) {
 	if file.relatedHandle == nil {
 		return newError("dbase-io-writememoheader-1", ErrNoFPT)
 	}
@@ -642,7 +647,7 @@ func (file *File) writeMemoHeader() (err error) {
 		return newError("dbase-io-writememoheader-4", err)
 	}
 	// Calculate the next free block
-	file.memoHeader.NextFree++
+	file.memoHeader.NextFree += uint32(size)
 	// Write the memo header
 	buf := make([]byte, 8)
 	binary.BigEndian.PutUint32(buf[:4], file.memoHeader.NextFree)

--- a/dbase/io_unix.go
+++ b/dbase/io_unix.go
@@ -537,9 +537,6 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 	if file.relatedHandle == nil {
 		return nil, newError("dbase-io-writememo-1", ErrNoFPT)
 	}
-	if length > int(file.memoHeader.BlockSize-8) {
-		return nil, newError("dbase-io-writememo-2", fmt.Errorf("memo data too large for block size %d", file.memoHeader.BlockSize))
-	}
 	// Get the block position
 	blockPosition := file.memoHeader.NextFree
 	// Write the memo header

--- a/dbase/io_unix.go
+++ b/dbase/io_unix.go
@@ -549,17 +549,17 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 		return nil, newError("dbase-io-writememo-2", err)
 	}
 	// Put the block data together
-	block := make([]byte, length+8)
+	data := make([]byte, 8)
 	// The first 4 bytes are the signature, 1 for text, 0 for binary(image)
 	if text {
-		binary.BigEndian.PutUint32(block[:4], 1)
+		binary.BigEndian.PutUint32(data[:4], 1)
 	} else {
-		binary.BigEndian.PutUint32(block[:4], 0)
+		binary.BigEndian.PutUint32(data[:4], 0)
 	}
 	// The next 4 bytes are the length of the data
-	binary.BigEndian.PutUint32(block[4:8], uint32(length))
+	binary.BigEndian.PutUint32(data[4:8], uint32(length))
 	// The rest is the data
-	copy(block[8:], raw)
+	data = append(data, raw...)
 	// Lock the block we are writing to
 	flock := &unix.Flock_t{
 		Type:   unix.F_WRLCK,
@@ -594,7 +594,7 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 		return nil, newError("dbase-io-writememo-5", err)
 	}
 	// Write the memo data
-	_, err = file.relatedHandle.Write(block)
+	_, err = file.relatedHandle.Write(data)
 	if err != nil {
 		return nil, newError("dbase-io-writememo-6", err)
 	}

--- a/dbase/io_windows.go
+++ b/dbase/io_windows.go
@@ -533,9 +533,6 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 	if file.relatedHandle == nil {
 		return nil, newError("dbase-io-writememo-1", ErrNoFPT)
 	}
-	if length > int(file.memoHeader.BlockSize-8) {
-		return nil, newError("dbase-io-writememo-2", fmt.Errorf("memo data too large for block size %d", file.memoHeader.BlockSize))
-	}
 	// Get the block position
 	blockPosition := file.memoHeader.NextFree
 	// Write the memo header
@@ -563,12 +560,12 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 		}
 		err = windows.LockFileEx(*file.relatedHandle, windows.LOCKFILE_EXCLUSIVE_LOCK, 0, blockPosition, blockPosition+uint32(file.memoHeader.BlockSize), o)
 		if err != nil {
-			return nil, newError("dbase-io-writememo-3", err)
+			return nil, newError("dbase-io-writememo-2", err)
 		}
 		defer func() {
 			ulockErr := windows.UnlockFileEx(*file.relatedHandle, 0, blockPosition, blockPosition+uint32(file.memoHeader.BlockSize), o)
 			if err != nil {
-				err = newError("dbase-io-writememoheader-4", ulockErr)
+				err = newError("dbase-io-writememoheader-3", ulockErr)
 			}
 		}()
 	}
@@ -577,17 +574,17 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 	// Seek to new the next free block
 	_, err = windows.Seek(*file.relatedHandle, position, 0)
 	if err != nil {
-		return nil, newError("dbase-io-writememo-5", err)
+		return nil, newError("dbase-io-writememo-4", err)
 	}
 	// Write the memo data
 	_, err = windows.Write(*file.relatedHandle, block)
 	if err != nil {
-		return nil, newError("dbase-io-writememo-6", err)
+		return nil, newError("dbase-io-writememo-5", err)
 	}
 	// Convert the block number to []byte
 	address, err := toBinary(blockPosition)
 	if err != nil {
-		return nil, newError("dbase-io-writememo-7", err)
+		return nil, newError("dbase-io-writememo-6", err)
 	}
 	return address, nil
 }

--- a/dbase/io_windows.go
+++ b/dbase/io_windows.go
@@ -545,17 +545,17 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 		return nil, newError("dbase-io-writememo-2", err)
 	}
 	// Put the block data together
-	block := make([]byte, length+8)
+	data := make([]byte, 8)
 	// The first 4 bytes are the signature, 1 for text, 0 for binary(image)
 	if text {
-		binary.BigEndian.PutUint32(block[:4], 1)
+		binary.BigEndian.PutUint32(data[:4], 1)
 	} else {
-		binary.BigEndian.PutUint32(block[:4], 0)
+		binary.BigEndian.PutUint32(data[:4], 0)
 	}
 	// The next 4 bytes are the length of the data
-	binary.BigEndian.PutUint32(block[4:8], uint32(length))
+	binary.BigEndian.PutUint32(data[4:8], uint32(length))
 	// The rest is the data
-	copy(block[8:], raw)
+	data = append(data, raw...)
 	// Lock the block we are writing to
 	if file.config.WriteLock {
 		o := &windows.Overlapped{
@@ -581,7 +581,7 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 		return nil, newError("dbase-io-writememo-4", err)
 	}
 	// Write the memo data
-	_, err = windows.Write(*file.relatedHandle, block)
+	_, err = windows.Write(*file.relatedHandle, data)
 	if err != nil {
 		return nil, newError("dbase-io-writememo-5", err)
 	}

--- a/dbase/io_windows.go
+++ b/dbase/io_windows.go
@@ -535,13 +535,17 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 	}
 	// Get the block position
 	blockPosition := file.memoHeader.NextFree
+	blocks := length / int(file.memoHeader.BlockSize)
+	if length%int(file.memoHeader.BlockSize) > 0 {
+		blocks++
+	}
 	// Write the memo header
-	err := file.writeMemoHeader()
+	err := file.writeMemoHeader(blocks)
 	if err != nil {
 		return nil, newError("dbase-io-writememo-2", err)
 	}
 	// Put the block data together
-	block := make([]byte, file.memoHeader.BlockSize)
+	block := make([]byte, length+8)
 	// The first 4 bytes are the signature, 1 for text, 0 for binary(image)
 	if text {
 		binary.BigEndian.PutUint32(block[:4], 1)
@@ -590,7 +594,8 @@ func (file *File) writeMemo(raw []byte, text bool, length int) ([]byte, error) {
 }
 
 // writeMemoHeader writes the memo header to the memo file.
-func (file *File) writeMemoHeader() (err error) {
+// Size is the number of blocks the new memo data will take up.
+func (file *File) writeMemoHeader(size int) (err error) {
 	if file.relatedHandle == nil {
 		return newError("dbase-io-writememoheader-1", ErrNoFPT)
 	}
@@ -618,7 +623,7 @@ func (file *File) writeMemoHeader() (err error) {
 		return newError("dbase-io-writememoheader-4", err)
 	}
 	// Calculate the next free block
-	file.memoHeader.NextFree++
+	file.memoHeader.NextFree += uint32(size)
 	// Write the memo header
 	buf := make([]byte, 8)
 	binary.BigEndian.PutUint32(buf[:4], file.memoHeader.NextFree)

--- a/dbase/table.go
+++ b/dbase/table.go
@@ -259,7 +259,7 @@ func New(version FileVersion, config *Config, columns []*Column, memoBlockSize u
 	}
 	// Write the memo header
 	if file.memoHeader != nil {
-		err = file.writeMemoHeader()
+		err = file.writeMemoHeader(0)
 		if err != nil {
 			return nil, err
 		}

